### PR TITLE
builtin_script_template_improvements

### PR DESCRIPTION
**Pull request by toger5**
~~# template~~
~~This adds another script template with a lot of additional information.~~
~~For the lazy users who just want to get a little idea without reading any docs.~~

~~Any thoughts? feel free to propose changes to the template.~~
### loading templates
the pr also changes how templates are loaded.
Before templates only where loaded when there is no script_template folder.
Now they are loaded when one template is missing in the script_template folder.

### CODE REVIEW

File: editor/editor_settings.cpp [hunk 1]
```diff
@@ -216,17 +216,83 @@ Variant _EDITOR_DEF(const String &p_var, const Variant &p_default) {
...
+			"	set_fixed_process(true) # This tells godot to call the '_fixed_process' callback on this node each PHYSICS frame\n"
+			"\n"
+			"	# THIS SECTION WILL ONLY WORK WHEN CREATING THE APPROPIATE NODES AND ATTACH SCRIPTS WITH THE CORRECT FUNCTIONS\n"
+			"	# Comment this section or add the needed nodes and adapt the names\n"
+			"	var my_node = $child_node_name # Accessing a child node and storing it in a variable. THERE HAS TO BE A CHILD NODE WITH THE NAME: 'child_node_name'\n"
+			"	var a_child_from_a_child = $child_node_name/sprite # Use '/' to access any node in the child tree from the current node.\n"
+			"	my_node.position = Vector2(100,100) # You can set any builtin parameter or call any function of the node you accessed.\n"
+			"	a_child_from_a_child.a_coustom_function(name) # You can call any function defined in the script attached to my_node just like that.\n"
+			"\n"
+			"_process(delta): \n"
```
Comment (Noshyaar) [inline editor/editor_settings.cpp]: Missing func

File: editor/editor_settings.cpp [hunk 2]
```diff
@@ -216,17 +216,83 @@ Variant _EDITOR_DEF(const String &p_var, const Variant &p_default) {
...
+			"func _fixed_process(delta):\n"
+			"	#this function is called each physics frame. And synced with the physics engine\n"
+			"	#the physics might be calculated in a different thread so interacting with it in _process might result in unwanted behaviour\n"
+			"	pass\n"
+			"\n"
+			"func _input(event):\n"
+			"	#the event variable can be of any Type that extends from 'InputEvent'"
+			"	if event is InputEventMouseButton: #Use the 'is' statement to check the type of the input event\n"
+			"		if event.button_index == BUTTON_LEFT: #each input event type has differnt variables. InputEventMouseButton has the parameter button_index.\n"
+			"			print('BUTTON_LEFT')\n"
```
Comment (Noshyaar) [inline editor/editor_settings.cpp]: Line 269-270 indentation issue

File: editor/editor_settings.cpp [hunk 3]
```diff
@@ -216,17 +216,83 @@ Variant _EDITOR_DEF(const String &p_var, const Variant &p_default) {
...
+			"	# Anything that needs to be calculated each frame goes here.\n"
+			"	pass\n"
+			"\n"
+			"func _fixed_process(delta):\n"
+			"	#this function is called each physics frame. And synced with the physics engine\n"
+			"	#the physics might be calculated in a different thread so interacting with it in _process might result in unwanted behaviour\n"
+			"	pass\n"
+			"\n"
+			"func _input(event):\n"
+			"	#the event variable can be of any Type that extends from 'InputEvent'"
```
Comment (Noshyaar) [inline editor/editor_settings.cpp]: Missing newline

File: editor/editor_settings.cpp [hunk 4]
```diff
@@ -216,17 +216,83 @@ Variant _EDITOR_DEF(const String &p_var, const Variant &p_default) {
...
+			"extends %BASE%\n"
+			"# ctrl+click or cmd+click to get information about the Node type this script is attached to.\n"
+			"# any variable and function is available in the scope of this script.\n"
+			"# don't forget to also checkout the classes this node Inherits from! All those functions + parameters are available too.\n"
+			"\n"
+			"\n"
+			"func _ready():\n"
+			"	# if you dont need one of the callbacks set it to false: set_process(false)"
+			"\n"
+			"	set_process(true) 		# This tells godot to call the '_ready' callback  on this node each frame\n"
```
Comment (Noshyaar) [inline editor/editor_settings.cpp]: `set_process*` is not necessary in 3.0. It will be enabled at the presence of _process function (and the likes).
The function can be useful for disabling & enabling it at runtime. Might be better to let the users know about this somehow.

(also double spaces after "callback")

File: editor/editor_settings.cpp [hunk 5]
```diff
@@ -216,17 +216,83 @@ Variant _EDITOR_DEF(const String &p_var, const Variant &p_default) {
...
+			"	pass\n"
+			"\n"
+			"func _input(event):\n"
+			"	#the event variable can be of any Type that extends from 'InputEvent'"
+			"	if event is InputEventMouseButton: #Use the 'is' statement to check the type of the input event\n"
+			"		if event.button_index == BUTTON_LEFT: #each input event type has differnt variables. InputEventMouseButton has the parameter button_index.\n"
+			"			print('BUTTON_LEFT')\n"
+			"	if event is InputEventKey: #To find the available parameters for a specific event: check the help page with: ctr+click on InputEventKey\n"
+			"		event.scancode == KEY_SPACE && event.pressed:\n"
+			"		print('KEY_SPACE')\n";
```
Comment (Noshyaar) [inline editor/editor_settings.cpp]: Missing 'if' and indentation

**Comment by Noshyaar** [comment]
Sorry for nitpicking. 😛  There are still some typos and discrepancies about whether to use `# comment`, `#comment`, `# Comment`, or `#Comment`.

I think this is only useful on the first use. To be honest, the current template, as minimal as it is, is kind of annoying to me because of its comments.

File: editor/editor_settings.cpp (hunk 5 above)
Comment (Hinsbart) [inline editor/editor_settings.cpp]: Also, please use %TS% for indentation, so it respects user settings regarding tabs vs spaces.

**Review by 27thLiz — DISMISSED** [review]

**Comment by Noshyaar** [comment]
In my opinion:

1. This is documentations' job. The docs might include some kind of "quick start to GDScript" similar to this.

2. Users are expected to read the docs. If it's tldr, the docs should be fixed.

3. GDScript is already pretty easy getting into.

Not saying that this *won't* be useful. It might be a good thing to give users general idea of how to program in GDScript, how to programmatically create an object, and what callbacks are available. However, I don't think the template is the right place. A first run reference (that's accessible later on) is more likely?

**Comment by toger5** [comment]
I still think such a template makes the experience a lot better. A lot of really new users just download and try around without even searching for docs. Just to get a feel for the application. In those szenarios it is great to have a template which gives u an overview.

**Comment by akien-mga** [comment]
I think that template is TL;DR, and as others mentioned, it's a bit weird to have a template that you would typically want to use only once. A documentation page showing the structure of a GDScript file with lots of comments and hints about available virtual methods would be much better IMO.

**Comment by toger5** [comment]
Do you still consider merging it for the second feature when i remove the additional template but leave the "loading when a template is missing instead of only loading templates when the whole script_template folder is missing"?

**Comment by toger5** [comment]
Also if anyone has an idea for an template which would be more useful pls share it.

**Comment by Hinsbart** [comment]
> Do you still consider merging it for the second feature when i remove the additional template but leave the "loading when a template is missing instead of only loading templates when the whole script_template folder is missing"?

Yeah, sounds good to me :)

**Comment by toger5** [comment]
remove the template.
I commented out so it is still there for reference if someone wants to improve it. Or has an idea for an additional template which is more useful.
Of course I'm also willing to completely remove it. Just tell me.

**Comment by akien-mga** [comment]
Please remove it completely, we don't need to introduce commented out code for a non-desired feature :)

Also your commit log should be in plain English, not in snake_case ;)

**Comment by toger5** [comment]
@akien-mga  done
